docs/v1/P4Runtime-Spec.mdk

```diff
@@ -3509,6 +3509,83 @@ gRPC provides utility functions `ExtractErrorDetails()` and `SetErrorDetails()`
 Please see sections on individual P4Runtime RPCs for details on how
 `grpc::Status` is populated for reporting errors.
 
+# Atomicity of Individual `Write` and `Read` Operations
+
+Each individual entity in a batch is guaranteed to be read or written atomically
+relative to packet forwarding. For example, for every table dataplane `apply`
+operation, and every single `Write` operation on a table that inserts, deletes,
+or modifies one table entry, the `apply` operation should behave as if that
+`Write` operation has not yet occurred, or as if the `Write` operation is
+complete. The P4 program should never behave as if the `Write` operation is
+partially complete. These guarantees also apply to extern instances: `Read` and
+`Write` operations on extern entities must execute atomically relative to extern
+dataplane methods.
+
+The atomicity guarantees provided by P4Runtime for individual `Read` and `Write`
+operations are the same as the guarantees required by PSA and are described in
+details in the PSA specification [@PSAAtomicityOfControlPlaneOps].
+
+The P4~16~ language introduces an `@atomic` annotation [@P4Concurrency], to
+guarantee atomic dataplane execution of entire blocks of P4 code. P4Runtime
+implementations are required to honor the `@atomic` annotation for `Write`
+operations, as well as [non-wildcard `Read` operations](#wildcard-reads),
+relative to dataplane execution. Consider the following P4 example written for
+PSA:
+
+~ Begin P4Example
+control C1 {
+  typedef bit<10> Index_t;
+  typedef bit<32> Value_t;
+  Register<Value_t, Index_t>(32w1024) r1;
+
+  apply {
+    // ...
+    @atomic {
+        Value_t v = r1.read((Index_t)1);
+        v = v + 1;
+        r1.write((Index_t)1, v);
+    }
+  }
+}
+~ End P4Example
+
+If a P4Runtime server is processing messages which write to Register `r1` at
+index `1`, these writes must not happen between the dataplane `read` and
+`write`.
+
+Now let's consider the following example:
+
+~ Begin P4Example
+control C1 {
+  typedef bit<10> Index_t;
+  typedef bit<32> Value_t;
+  Register<Value_t, Index_t>(32w1024, (Value_t)0 /* initial value */) r1;
+
+  apply {
+    @atomic {
+        r1.write((Index_t)1, (Value_t)100);
+        r1.write((Index_t)2, (Value_t)100);
+    }
+    @atomic {
+        r1.write((Index_t)1, (Value_t)200);
+        r1.write((Index_t)2, (Value_t)200);
+    }
+  }
+}
+~ End P4Example
+
+If a P4Runtime client issues a *wildcard* `Read` on Register `r1`, there is no
+guarantee that `r1[1] == r1[2]` in the response, as the read for `r1[1]` and the
+read for `r1[2]` may not happen "within" the same atomic block. The server is
```

Comment:
suggested change, if it matches what you intended to mean:

Replace this: 'may not happen "within" the same atomic block.'

with this: 'may happen after the data plane executes the first atomic block, but before it executes the second atomic block.'

Reply:
or maybe even more precisely (I think I messed up with my previous suggestion):

replace the phrase starting with 'as the read ...' with this instead: 'as the read for `r1[1]` may occur after the data plane executes the first atomic block, but before the second atomic block, and the read for `r1[2] may occur after the data plane executes the second atomic block.'

Reply:
I made that change. Thanks!